include/particles/particle_xmpm.tcc

```diff
@@ -0,0 +1,266 @@
+
+  // Assign volumetric strain at centroid
+  dvolumetric_strain_ = dt * strain_rate_centroid.head(Tdim).sum();
+  volumetric_strain_centroid_ += dvolumetric_strain_;
+}
+
+//! Map body force
+template <unsigned Tdim>
+void mpm::ParticleXMPM<Tdim>::map_body_force(
+    const VectorDim& pgravity) noexcept {
+  // Compute nodal body forces
+  for (unsigned i = 0; i < nodes_.size(); ++i) {
+    nodes_[i]->update_external_force(true, mpm::ParticlePhase::Solid,
+                                     (pgravity * mass_ * shapefn_(i)));
+    if (nodes_[i]->discontinuity_enrich())
+      nodes_[i]->update_discontinuity_property(
+          true, "external_force_enrich",
+          sgn(levelset_phi_) * pgravity * mass_ * shapefn_(i), 0, Tdim);
+  }
+}
+
+//! Map internal force
+template <>
+inline void mpm::ParticleXMPM<1>::map_internal_force() noexcept {
+  // Compute nodal internal forces
+  for (unsigned i = 0; i < nodes_.size(); ++i) {
+    // Compute force: -pstress * volume
+    Eigen::Matrix<double, 1, 1> force;
+    force[0] = -1. * dn_dx_(i, 0) * volume_ * stress_[0];
+
+    nodes_[i]->update_internal_force(true, mpm::ParticlePhase::Solid, force);
+    if (nodes_[i]->discontinuity_enrich())
+      nodes_[i]->update_discontinuity_property(
+          true, "internal_force_enrich", sgn(levelset_phi_) * force, 0, 1);
+  }
+}
+
+//! Map internal force
+template <>
+inline void mpm::ParticleXMPM<2>::map_internal_force() noexcept {
+  // Compute nodal internal forces
+  for (unsigned i = 0; i < nodes_.size(); ++i) {
+    // Compute force: -pstress * volume
+    Eigen::Matrix<double, 2, 1> force;
+    force[0] = dn_dx_(i, 0) * stress_[0] + dn_dx_(i, 1) * stress_[3];
+    force[1] = dn_dx_(i, 1) * stress_[1] + dn_dx_(i, 0) * stress_[3];
+
+    force *= -1. * this->volume_;
+
+    nodes_[i]->update_internal_force(true, mpm::ParticlePhase::Solid, force);
+    if (nodes_[i]->discontinuity_enrich())
+      nodes_[i]->update_discontinuity_property(
+          true, "internal_force_enrich", sgn(levelset_phi_) * force, 0, 2);
+  }
+}
+
+//! Map internal force
+template <>
+inline void mpm::ParticleXMPM<3>::map_internal_force() noexcept {
+  // Compute nodal internal forces
+  for (unsigned i = 0; i < nodes_.size(); ++i) {
+    // Compute force: -pstress * volume
+    Eigen::Matrix<double, 3, 1> force;
+    force[0] = dn_dx_(i, 0) * stress_[0] + dn_dx_(i, 1) * stress_[3] +
+               dn_dx_(i, 2) * stress_[5];
+
+    force[1] = dn_dx_(i, 1) * stress_[1] + dn_dx_(i, 0) * stress_[3] +
+               dn_dx_(i, 2) * stress_[4];
+
+    force[2] = dn_dx_(i, 2) * stress_[2] + dn_dx_(i, 1) * stress_[4] +
+               dn_dx_(i, 0) * stress_[5];
+
+    force *= -1. * this->volume_;
+
+    nodes_[i]->update_internal_force(true, mpm::ParticlePhase::Solid, force);
+    if (nodes_[i]->discontinuity_enrich())
+      nodes_[i]->update_discontinuity_property(
+          true, "internal_force_enrich", sgn(levelset_phi_) * force, 0, 3);
+  }
```

Comment:
Rather than rewriting the whole function, which is a code duplicacy, why don't we call the base class function `mpm::Particle<3>::map_internal_force()`?
```suggestion
  // Compute nodal internal forces
  mpm::Particle<3>::map_internal_force();
  // Compute nodal internal forces enrich
  for (unsigned i = 0; i < nodes_.size(); ++i) {
    if (nodes_[i]->discontinuity_enrich())
      nodes_[i]->update_discontinuity_property(
          true, "internal_force_enrich", sgn(levelset_phi_) * force, 0, 3);
  }
```

Reply:
By doing that we need to compute the force twice.